### ISSUE TITLE
OGRE doesn't display meshes anymore

### ISSUE THREAD

**Issue by rhaschke** [microsoft/wslg#744]
I'm maintaining an OpenGL application, called rviz, that doesn't correctly display meshes since a few weeks: https://github.com/ros-visualization/rviz/issues/1748

Simple geometries are correctly rendered. As the program runs fine on Linux natively and we didn't changed the rendering pipeline recently, I'm afraid this issue is related to some recent change in WSLg or Mesa.

Looks like WSLg forwards OpenGL via the Mesa driver to the NVIDIA card. I tried building against both, Ogre 1.9 (default) and Ogre 1.12 (which is the underlying rendering library). Both fail to display meshes. Using another application, `meshlab`, the meshes render correctly.

Was there any change released recently that could explain this regression?

# Environment

I need to check these details next time I'm booting into Windows...
```none
Windows build number: [run `[Environment]::OSVersion` for powershell, or `ver` for cmd]
Your Distribution version: Ubuntu 20.04
Your WSLg version: [Open 'Apps and Features' and check the version of 'Windows Subsystem for Linux GUI app support', e.g: 0.2.3.13]
```

**WSL logs**: 

* Attach WSLg logs from  `/mnt/wslg`

You can access the wslg logs using explorer at: `\\wsl$\<Distro-Name>\mnt\wslg` (e.g.: `\\wsl$\Ubuntu-20.04\mnt\wslg`)

* `pulseaudio.log`
* `weston.log`
* `versions.txt`

**Comment by rhaschke** [comment]
On another laptop, having an Intel GPU only, the meshes are displayed correctly. Here OpenGL (glxinfo) reports:
```
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: llvmpipe (LLVM 12.0.0, 256 bits)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 21.2.6
```
I conclude that the issue is related to Microsoft's D3D12 renderer.

**Comment by rhaschke** [comment]
Is it possible to resort to the Intel GPU and/or the Mesa software renderer to workaround this issue?

**Comment by IoTDan** [comment]
I have this happening with an Intel Iris XE graphics card, with the MESA drivers installed. setting LIBGL_ALWAYS_SOFTWARE=true does correctly render. So looking for a hardware accelerated fix.

**Comment by hideyukn88** [comment]
@IoTDan, would you please run `glxgears` on your system? Does it crash? We are aware of GL app crash issue with Intel XE graphics, and wondering if your case is same, please let us know, thanks!

**Comment by IoTDan** [comment]
I did the other day. Runs doesn’t crash.

Thanks,

Dan
________________________________

**Comment by rhaschke** [comment]
@hideyukn88, I don't think this is only related to Intel GPUs, but to `D3D12` in general. Above, I stated that WSLg picks my NVIDIA card and the D3D12 driver showing these problems (as well).

**Comment by rhaschke** [comment]
I have a Laptop with an integrated `Intel UHD` GPU _and_ an `NVIDIA GeForce MX250`.
| software rendering via llvmpipe | accelerated rendering via D3D12 |
|--------|------------|
| ![image](https://user-images.githubusercontent.com/5376030/185242913-5fc70506-eab1-4073-a83e-24b2f79ec0f1.png) | ![image](https://user-images.githubusercontent.com/5376030/185243049-ec816cfd-33f0-4a2d-9014-8853c8ddbba4.png) |
| OpenGL device: llvmpipe (LLVM 12.0.0, 256 bits) | OpenGL device: D3D12 (NVIDIA GeForce MX250) |
| OpenGl version: 3.1 (GLSL 1.4) | OpenGl version: 3.1 (GLSL 1.4) |

I don't know how to switch to the Intel GPU to report results for it as well. Is there an environment variable to set?

**Comment by hideyukn88** [comment]
@rhaschke, please refer below for switching GPUs, and if it reproduces with NVIDIA, would you please share the driver version, thanks!

https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10710

**Comment by rhaschke** [comment]
Switching the GPU via the env variable `MESA_D3D12_ADAPTER_NAME=intel` doesn't work for me. `glxinfo` always reports the NVIDIA GPU being selected.

NVIDIA driver: 30.0.15.1236
Intel driver: 30.0.101.1338

Looks like my WSL wasn't updated for a while. However, Windows Update doesn't report any pending updates.
```
> ver
Microsoft Windows [Version 10.0.22000.856]

> wsl --status
Default Distribution: Ubuntu
Default Version: 2

Windows Subsystem for Linux was last updated on 26.05.2022
WSL automatic updates are on.

Kernel version: 5.10.102.1
```

**Comment by jorolf** [comment]
> Switching the GPU via the env variable `MESA_D3D12_ADAPTER_NAME=intel` doesn't work for me

They changed the variable name from `MESA_D3D12_ADAPTER_NAME` to `MESA_D3D12_DEFAULT_ADAPTER_NAME` before merging the merge request. (See also: https://docs.mesa3d.org/drivers/d3d12.html#envvar-MESA_D3D12_DEFAULT_ADAPTER_NAME)

**Comment by xiayip** [comment]
@jorolf 
Use MESA_D3D12_DEFAULT_ADAPTER_NAME, the GPU switched, but the problem still exist. I think its the problem with D3D12, not GPU hardware.
